### DISCUSSION TITLE
Failed to build older Ruby versions on Catalina

### DISCUSSION THREAD

**Discussion by coding-red-panda** [rbenv/ruby-build#1852] [Build failures] [answered]
Hello,

since the last update to XCode and Command Line Tools, which puts you on `clang 12.0`, it's impossible to build older Ruby versions anymore. Every time I try to install a version, I can see that ruby-build downloads the older OpenSSL manually and applies the patches, but in the end the compilation keeps failing on the system.
So far the only solution is to downgrade Command Line Tools, but this prevents gems like `mysql2` from being compiled then....

Here's the log output:

```
 rbenv install 2.1.10
Downloading openssl-1.0.2u.tar.gz...
-> https://dqw8nmjcqpjn7.cloudfront.net/ecd0c6ffb493dd06707d38b14bb4d8c2288bb7033735606569d8f90f89669d16
Installing openssl-1.0.2u...
Installed openssl-1.0.2u to /Users/olivar/.rbenv/versions/2.1.10

Downloading ruby-2.1.10.tar.bz2...
-> https://cache.ruby-lang.org/pub/ruby/2.1/ruby-2.1.10.tar.bz2
Installing ruby-2.1.10...

WARNING: ruby-2.1.10 is past its end of life and is now unsupported.
It no longer receives bug fixes or critical security updates.

ruby-build: using readline from homebrew

BUILD FAILED (Mac OS X 10.15.7 using ruby-build 20200926)

Inspect or clean up the working tree at /var/folders/n5/t6kf526x431dsyk85m8v7slh0000gn/T/ruby-build.20201008093719.18645.wHoGzN
Results logged to /var/folders/n5/t6kf526x431dsyk85m8v7slh0000gn/T/ruby-build.20201008093719.18645.log

Last 10 log lines:
linking shared-object racc/cparse.bundle
linking shared-object json/ext/generator.bundle
installing default psych libraries
6 warnings generated.
linking shared-object bigdecimal.bundle
linking shared-object psych.bundle
linking shared-object nkf.bundle
13 warnings generated.
linking shared-object date_core.bundle
make: *** [build-ext] Error 2
```

### Log file output
```
make[2]: Nothing to be done for `all'.
In file included from ossl_ns_spki.c:11:
In file included from ./ossl.h:219:
./openssl_missing.h:71:6: error: conflicting types for 'HMAC_CTX_copy'
void HMAC_CTX_copy(HMAC_CTX *out, HMAC_CTX *in);
     ^
/Users/olivar/.rbenv/versions/2.1.10/openssl/include/openssl/hmac.h:101:5: note: previous declaration is here
int HMAC_CTX_copy(HMAC_CTX *dctx, HMAC_CTX *sctx);
    ^
In file included from ossl_ns_spki.c:11:
In file included from ./ossl.h:219:
./openssl_missing.h:95:5: error: conflicting types for 'EVP_CIPHER_CTX_copy'
int EVP_CIPHER_CTX_copy(EVP_CIPHER_CTX *out, EVP_CIPHER_CTX *in);
    ^
/Users/olivar/.rbenv/versions/2.1.10/openssl/include/openssl/evp.h:542:5: note: previous declaration is here
int EVP_CIPHER_CTX_copy(EVP_CIPHER_CTX *out, const EVP_CIPHER_CTX *in);
    ^
In file included from ossl_ns_spki.c:11:
In file included from ./ossl.h:219:
./openssl_missing.h:173:5: error: conflicting types for 'BN_rand_range'
int BN_rand_range(BIGNUM *r, BIGNUM *range);
    ^
/Users/olivar/.rbenv/versions/2.1.10/openssl/include/openssl/bn.h:484:5: note: previous declaration is here
int BN_rand_range(BIGNUM *rnd, const BIGNUM *range);
    ^
In file included from ossl_ns_spki.c:11:
In file included from ./ossl.h:219:
./openssl_missing.h:177:5: error: conflicting types for 'BN_pseudo_rand_range'
int BN_pseudo_rand_range(BIGNUM *r, BIGNUM *range);
    ^
/Users/olivar/.rbenv/versions/2.1.10/openssl/include/openssl/bn.h:485:5: note: previous declaration is here
int BN_pseudo_rand_range(BIGNUM *rnd, const BIGNUM *range);
    ^
In file included from ossl_ns_spki.c:11:
In file included from ./ossl.h:233:
./ossl_pkey.h:43:5: warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
#if HAVE_BN_GENCB
    ^
./ossl_pkey.h:42:23: note: expanded from macro 'HAVE_BN_GENCB'
#define HAVE_BN_GENCB defined(HAVE_RSA_GENERATE_KEY_EX) || defined(HAVE_DH_GENERATE_PARAMETERS_EX) || defined(HAVE_DSA_GENERATE_PARAMETERS_EX)
                      ^
./ossl_pkey.h:43:5: warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
./ossl_pkey.h:42:60: note: expanded from macro 'HAVE_BN_GENCB'
#define HAVE_BN_GENCB defined(HAVE_RSA_GENERATE_KEY_EX) || defined(HAVE_DH_GENERATE_PARAMETERS_EX) || defined(HAVE_DSA_GENERATE_PARAMETERS_EX)
                                                           ^
./ossl_pkey.h:43:5: warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
./ossl_pkey.h:42:103: note: expanded from macro 'HAVE_BN_GENCB'
#define HAVE_BN_GENCB defined(HAVE_RSA_GENERATE_KEY_EX) || defined(HAVE_DH_GENERATE_PARAMETERS_EX) || defined(HAVE_DSA_GENERATE_PARAMETERS_EX)
                                                                                                      ^
3 warnings and 4 errors generated.
compiling psych.c
compiling sizes.c
make[2]: *** [ossl_ns_spki.o] Error 1
make[1]: *** [ext/openssl/all] Error 2
make[1]: *** Waiting for unfinished jobs....
installing default cparse libraries
linking shared-object pty.bundle
compiling psych_parser.c
installing default sizeof libraries
compiling ../.././ext/psych/yaml/loader.c
compiling ../.././ext/psych/yaml/scanner.c
linking shared-object objspace.bundle
linking shared-object rbconfig/sizeof.bundle
compiling ../.././ext/psych/yaml/api.c
compiling ../.././ext/psych/yaml/writer.c
compiling ../.././ext/psych/yaml/dumper.c
linking shared-object json/ext/parser.bundle
compiling ../.././ext/psych/yaml/reader.c
compiling ../.././ext/psych/yaml/emitter.c
2 warnings generated.
linking shared-object pathname.bundle
compiling ../.././ext/psych/yaml/parser.c
linking shared-object racc/cparse.bundle
linking shared-object json/ext/generator.bundle
installing default psych libraries
6 warnings generated.
linking shared-object bigdecimal.bundle
linking shared-object psych.bundle
linking shared-object nkf.bundle
13 warnings generated.
linking shared-object date_core.bundle
make: *** [build-ext] Error 2
```

**Comment by ricardoham** [comment]
Same issue here
macOS 10.15.7
ruby 2.6.3p62(current version on machine)
rbenv 1.1.2
ruby-build 20201005

I tried a lot of alternatives as follows: 

via openssl1.0 and openssl > 1.0(same result)
RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.0)" rbenv install 2.1.0
result: 
The Ruby zlib extension was not compiled.
ERROR: Ruby install aborted due to missing extensions

via zlib
LDFLAGS=-L/usr/local/opt/zlib/lib CPPFLAGS=-I/usr/local/opt/zlib/include rbenv install 2.1.0
result:
BUILD FAILED (Mac OS X 10.15.7 using ruby-build 20201005)

**Comment by microspino** [comment]
Same here with OS X 10.15.7. I'm unable to install any 2.2.x version of ruby.
Darwin ... 19.6.0 Darwin Kernel Version 19.6.0: Mon Aug 31 22:12:52 PDT 2020; 
root:xnu-6153.141.2~1/RELEASE_X86_64 x86_64

I've install openssl@1 with homebrew. This is the output of one of my attempts:

skip installing bundle gems because of lacking zlib
installing rdoc:              /Users/../.rbenv/versions/2.2.2/share/ri/2.2.0/system
installing capi-docs:         /Users/../.rbenv/versions/2.2.2/share/doc/ruby
The Ruby openssl extension was not compiled.
The Ruby readline extension was not compiled.
The Ruby zlib extension was not compiled.
ERROR: Ruby install aborted due to missing extensions
Configure options used:
  --prefix=/Users/.../.rbenv/versions/2.2.2
  --with-openssl-dir=/usr/local/opt/openssl@1.0
  --with-gcc=clang
  --with-readline-dir=/usr/local/opt/readline
  CC=clang
  LDFLAGS=-L/Users/../.rbenv/versions/2.2.2/lib -L/usr/local/opt/openssl@1.0/lib
  CPPFLAGS=-I/Users/.../.rbenv/versions/2.2.2/include -I/usr/local/opt/openssl@1.0/include

Please note that **installing 2.3.8 worked fine.**

**I can confirm @coding-bunny assumption: using the Command_Line_Tools_for_Xcode_11.4.1 WORKS** (I had to remove Xcode 12 but maybe you don't use it).

**Comment by coding-red-panda** [comment]
Using however the Command Line Tools for XCode v11 on a Cataline System with XCode 12 installed introduces other problems however. You'll notice that you won't be able to compile the mysql2 gem for example unless you reset your `xcode-select` to the default XCode path and them run `bundle install`.

Even if you use `bundle config` and manually supply all paths, with the above description mysql2 gem will refuse to compile.
It's kind of a catch-22 : Need the V11 to compile rubies, but the V12 to compile gems....

**Comment by microspino** [comment]
**coding-bunny** you can however do the opposite procedure after compiling ruby versions.
I mean reverting back to v 12 after running rbenv install. Am I wrong?

**Comment by coding-red-panda** [comment]
I can, but I shouldn't have to....
Working in an environment where I need to swap between over 12 different ruby versions, and go through this entire hassle every single time someone gets a new macbook or joins the company is simply not sustainable.

`ruby-build` and the relevant gems should compile regardless of whether V11 or V12 is used/installed.

**Comment by eddroid** [comment]
Sounds like https://github.com/rbenv/ruby-build/issues/1489

**Comment by coding-red-panda** [comment]
> Sounds like #1489

If Apple indeed changed/removed underlying CFlags from the compiler, then yes it could very well be the main cause why stuff stopped working. The question is whether ruby-build is going to be able to work around this...

**Comment by ricardoham** [comment]
> Sounds like #1489

This worked for me on 10.15.7 macOS
Thank you

**Comment by rohitnand10** [comment]
#1489 Did not work for me and the errors I get are:
Last 10 log lines:
./ossl_pkey.h:43:103: note: expanded from macro 'HAVE_BN_GENCB'
#define HAVE_BN_GENCB defined(HAVE_RSA_GENERATE_KEY_EX) || defined(HAVE_DH_GENERATE_PARAMETERS_EX) || defined(HAVE_DSA_GENERATE_PARAMETERS_EX)
                                                                                                      ^
5 warnings and 4 errors generated.
make[2]: *** [ossl_pkcs7.o] Error 1
make[1]: *** [ext/openssl/all] Error 2
make[1]: *** Waiting for unfinished jobs....
linking shared-object objspace.bundle
linking shared-object nkf.bundle
make: *** [build-ext] Error 2

**Comment by franck-nadeau** [comment]
I think this is an issue with M1 arch. I had the same issue and resolved it by using the following:

````
 arch -x86_64 rbenv install 2.7.6
````

**Comment by mislav** [comment, marked as answer]
Answering for posterity, since people keep discovering these outdated threads in search results:

As macOS and its Command Line Tools get upgrades, older versions of Ruby will become harder or impossible to build. This is just the state of the world that we live in. Compatibility is further complicated by newer Mac machines having moved to Apple Silicon architecture, where it is even less likely to build an older version of Ruby. **The solution: update Ruby in your application.**

Some people discover and share workarounds such as setting arcane CFLAGS settings or using `arch -x86_64` to trick things to be compiled for `amd64` architecture even while on `arm64`, but these workarounds are more trouble than they're worth, and will definitely break down or cause troubles over time.

Upgrade your Ruby version to a currently maintained Ruby release or, if that is not an option for your app, create a development environment for your app within a virtual machine or a container.
